account_bank_statement_import_online/models/online_bank_statement_provider.py

```diff
@@ -141,184 +118,237 @@ def _compute_name(self):
+        statement = AccountBankStatement.search(domain, limit=1)
+        if statement:
+            statement.write(statement_values)
+        else:
+            statement_values.update(
+                {
+                    "name": self.journal_id.sequence_id.with_context(
+                        ir_sequence_date=statement_date,
+                    ).next_by_id(),
+                    "journal_id": self.journal_id.id,
+                    "date": statement_date,
+                }
+            )
+            statement = AccountBankStatement.with_context(
+                journal_id=self.journal_id.id,
+            ).create(statement_values)
 
-                    bank_account_number = line_values.get("account_number")
-                    if bank_account_number:
-                        line_values.update(
-                            {
-                                "account_number": (
-                                    self._sanitize_bank_account_number(
-                                        bank_account_number
-                                    )
-                                ),
-                            }
-                        )
+    def _update_statement_balance_values(
+        self,
+        statement_values,
+        lines_data,
+        statement_date_since,
+        statement_date_until
+    ):
+        """Update statement start or end balance from line data.
 
-                    filtered_lines.append(line_values)
-                statement_values.update(
-                    {
-                        "line_ids": [[0, False, line] for line in filtered_lines],
-                    }
-                )
+        Statements retrieved and statements created might have different begin
+        and end dates. Suppose we retrieve data that covers three days, and has four
+        transactions, but the statement to be created is for only one day. like this:
+        statement_values: {
+            "balance_start": 25.00,
+            "balance_end_real": 150.00,
+        }
+        lines_data = [
+            {"date": "2020-02-01", "amount": 25.00},
+            {"date": "2020-02-02", "amount": 25.00},
+            {"date": "2020-02-02", "amount": 50.00},
+            {"date": "2020-02-03", "amount": 25.00},
+        ]
+        If we create a statement for only the day 2020-02-02, the first and
+        the last line should not belong to this statement. But at the start of
+        2020-02-02 the balance was 50.00. So the amounts for lesser dates should
+        be added to the balance_start. Similarly amounts from later dates should
+        be subtracted from balance_end_real. Which should be 125.00 for
+        2020-02-02.
+        """
+        for line_values in lines_data:
+            date = self._get_utc_line_date(line_values)
+            date = date.astimezone(utc).replace(tzinfo=None)
+            # Add amounts from dates before statement to balance_start.
+            if date < statement_date_since:
```

Comment:
Is this correct, timezone-wise? If I have a statement from '2020-02-03 00:01', CEST, then in  UTC it will be yesterday 22:01, and it will be lower than the statement_date_since which is `2020-02-04`, so it will be counted as the day before, but that's not correct.

Shouldn't `get_utc_line_date` be replaced with a function `get_line_date` which always returns a naive date without a time component, as the date was in the Bank's timezone, which can then be used to determine which date the statement should get?

Reply:
I tried to keep the timezone logic as it was before.

However correct handling of timezones is quite difficult:,
All datetime fields in the database are assumed to be in UTC format.
The bank can be ine one timezone,
The server can be in a second timezone,
The user can be in a third timezone
The company holding the accounts can be in a fourth.

Best to convert all incoming datetimes as soon as [possible to UTC.

However when considering which date (as opposed to datetime) to use for the bank statement, we have IMHO to look at the timezone for the company that holds the account. The res.company does not have that atttribute at the moment, so we would need to add it.

Or we can simply use the timezone of the bank, if this differs from the timezone of the company, the meaning of the statement date will be the date in the timezone of the bank. The datetime of the line will still need to be stored as utc.

Will check the code for this.

Reply:
Instead of making guesses which is better, why not just to let (which was already a feature IIRC) user configure which TZ to convert to?

Reply:
@NL66278 Yes we should convert to UTC for datetimes, but bank statements are tied to dates, so we have to make a choice somewhere. Configurable is a good idea. I don't know where that was implemented but since @alexey-pelykh is the original author it must be somewhere :-) My guess would be on the provider record.

Reply:
I will add a configuration option then.

Reply:
Isn't it [this one](https://github.com/OCA/bank-statement-import/blob/14.0/account_statement_import_online/models/online_bank_statement_provider.py#L39)

Reply:
Or i mean, [this one](https://github.com/OCA/bank-statement-import/blob/12.0/account_bank_statement_import_online/models/online_bank_statement_provider.py#L49), in 12.0

Reply:
The thing is, we need two timezone fields. One for the timezone of the bank. One fr the timezone of the company (to be set on the provider).  Unless the dates in the data from the bank are timezone aware.

Reply:
Ahhh yes, you're right. So I guess it needs `source_tz` and `tz` (the latter being `dest_tz`)

Reply:
On second thought: is that really needed? Because if we get statements like: 2022-03-01 00:30, and we say that this should be interpreted as the timezone GMT+3 where the bank is in; then we save the statement lines as the UTC equivalent and they will appear at the right date and time in the user's timezone; and for the statement date we use the bank's timezone and truncate the time. Why do we then still need the company timezone?

Reply:
Well if we have a transaction for a company that occuring to its utc timestamp happened 2020-02-01 at 23:00, this  means for the company it occured on 2020-02-02 at 01:00, so should go into a statement for 2020-02-02. This problem becomes bigger the further companies / Odoo customers are from UTC.

Reply:
But that's assuming the company's timezone determines the date of the transactions, whereas I would think the timezone of the bank decides that.

I mean, if you have a bank account in hong kong, the transactions will be divided in statements per day according to the hong kong timezone, right?

Reply:
Wrong. The division of transactions within Odoo over statements (daily, weekly or monthly) has no relation at all with the organization of transactions at the online provider / bank. Actually, the provider might not have organized the transactions into statements at all. (For instance Ponto has no concept of statements). The transactions retrieved might have timezone data in their dates. Or not, then we need a default for the provider. For instande Qonto is supposed to be in Paris. We have to translate the transaction timestamps from the bank, into timestamps for the company / Odoo customer, to determine to which day, week or month they belong.